content/blog/how-to-test-react-app-with-node-test-runner.en.mdx

```diff
@@ -0,0 +1,164 @@
+---
+title: How to test a React app with Node Test Runner
+description: You ever wanted to test your React app with Node Test Runner? Here's how you can do it.
+date: 2024-11-19
+authors: AugustinMauroy
+---
+
+## Introduction
+
+Testing is a crucial part of software development, ensuring that your application behaves as expected. With the [Node.js Test Runner](https://nodejs.org/en/learn/test-runner/introduction), you can seamlessly integrate testing into your [React](https://react.dev) application. This guide will walk you through the process of setting up and running tests for a React app using the Node.js Test Runner.
+
+> I really recommend you to read the [Node.js Test Runner documentation](https://nodejs.org/en/learn/test-runner/introduction) to understand how node.js test runner works.
+> This guide assumes you have a basic understanding of React and testing.
+
+## Installing the Dependencies
+
+First, you need to install the necessary dependencies. In addition to your React app dependencies, you will need the following:
+
+```bash
+npm install --save-dev @testing-library/react @testing-library/dom jsdom
+```
+
+> **Note**: The rest of the dependencies we will use come from your runtime (Node.js).
+
+## Writing the Component to Be Tested
+
+Let's create a simple React component that we will test. This component will be a counter that increments a value when a button is clicked.
+
+```tsx fileName="src/components/Counter/index.tsx"
+'use client';
+import { useState } from 'react';
+import styles from './index.module.css';
+import type { FC } from 'react';
+
+const Counter: FC = () => {
+  const [count, setCount] = useState(0);
+
+  return (
+    <div className={styles.container}>
+      <p className={styles.count}>{count}</p>
+      <button onClick={() => setCount(count + 1)} className={styles.button}>
+        Increment
+      </button>
+    </div>
+  );
+};
+
+export default Counter;
+```
+
+<details>
+  <summary>`index.module.css`</summary>
+
+  ```css fileName="src/components/Counter/index.module.css"
+  .container {
+    @apply flex flex-col items-center justify-center;
+
+    .count {
+      @apply text-4xl;
+    }
+
+    .button {
+      @apply px-4 py-2 bg-blue-500 text-white rounded-md;
+    }
+  }
+  ```
+</details>
+
+## Registering Node.js Loaders
+
+To handle TypeScript and CSS modules, you need to register the appropriate loaders. Create a file named `node-hooks/react-test.js` and add the following code:
+
+> If you want to understand what is a loader. You need to read [this post](/blog/post/how-to-use-nodejs-loader) I had written about loaders.
+
+```mjs fileName="node-hooks/react-test.js"
+import { register } from 'node:module';
+import { pathToFileURL } from 'node:url';
+import jsdom from 'jsdom';
+
+// Register the loaders
+register('nodejs-loaders/dev/tsx', pathToFileURL('./'));
+register('nodejs-loaders/testing/css-module', pathToFileURL('./'));
```

Comment:
v2 will be published imminently, so I would just publish your article assuming v2

I think you don't need the 2nd argument for register here because they're node_modules

```suggestion
register('@nodejs-loaders/tsx');
register('@nodejs-loaders/css-module');
```

Reply:
I just tested on my blog and we need it

Reply:
Hmm. Interesting